qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -912,23 +913,220 @@ def __call__(self, target, basis_fidelity=None, *, _num_basis_uses=None) -> Quan
         if _num_basis_uses is not None:
             best_nbasis = _num_basis_uses
         decomposition = self.decomposition_fns[best_nbasis](target_decomposed)
-        decomposition_euler = [self._decomposer1q._decompose(x) for x in decomposition]
-
-        q = QuantumRegister(2)
-        return_circuit = QuantumCircuit(q)
-        return_circuit.global_phase = target_decomposed.global_phase
-        return_circuit.global_phase -= best_nbasis * self.basis.global_phase
-        if best_nbasis == 2:
-            return_circuit.global_phase += np.pi
-        for i in range(best_nbasis):
-            return_circuit.compose(decomposition_euler[2 * i], [q[0]], inplace=True)
-            return_circuit.compose(decomposition_euler[2 * i + 1], [q[1]], inplace=True)
-            return_circuit.append(self.gate, [q[0], q[1]])
-        return_circuit.compose(decomposition_euler[2 * best_nbasis], [q[0]], inplace=True)
-        return_circuit.compose(decomposition_euler[2 * best_nbasis + 1], [q[1]], inplace=True)
 
+        if isinstance(self.gate, CXGate):
+            num_u = 2 * best_nbasis
+            decomposition_euler = [None] * num_u
+            for i in range(0, num_u, 2):
+                pass
+        else:
+            decomposition_euler = [self._decomposer1q._decompose(x) for x in decomposition]
+        if self.pulse_optimize and best_nbasis in {2, 3}:
+            if self._decomposer1q.basis in {"ZSX", "ZSXX"}:
+                if isinstance(self.gate, CXGate):
+                    if best_nbasis == 3:
+                        return_circuit = self._get_sx_vz_3cx_efficient_euler(
+                            decomposition, target_decomposed
+                        )
+                    elif best_nbasis == 2:
+                        return_circuit = self._get_sx_vz_2cx_efficient_euler(
+                            decomposition, target_decomposed
+                        )
+                else:
+                    raise QiskitError(
+                        "pulse_optimizer currently only works with CNOT " "entangling gate"
+                    )
+            else:
+                raise QiskitError(
```

Comment:
Since the `pulse_optimizer` setting is default False, from the perspective of this class it must be intentionally set, which may mean that not synthesizing  "pulse optimally" might be seen as unexpected. I realize, however, that from the perspective of a transpiler pass this intention may be hidden but then it may be the responsibility of the caller to catch the exception and behave accordingly (e.g. call with `pulse_optimize=False`).

Reply:
Lev suggested an interface which may handle user intention; `pulse_optimize` can have three possible values:
None: do pulse optimization if available but don't error/warn if not
True: do pulse optimization and raise if not available
False: don't do optimization.
Having None as default then gives best circuit even when someone isn't familiar with the setting.